File: 2018.8.5/Vjudge/Knapsack.py
# Knapsack.py

def f(i, C):
    if C < 0: return 0
    if i < 0: return 0
    if C - w[i] >= 0:
        return max(f(i - 1, C), f(i - 1, C - w[i]) +v[i])
    return f(i - 1, C)

File: 2018.8.5/Vjudge/Knapsack.cpp
#include<stdio.h>
int max(int a, int b)
{
	return a > b ? a : b;
}

int solve(int C, int i, int* s, int* v)
{
	if (C < 0 || i < 0) return 0;
	if(C - s[i] >= 0)
		return max(solve(C - s[i], i - 1, s, v) + v[i], solve(C, i - 1, s, v));
	return solve(C, i - 1, s, v);
}

int main()
{
	int S, N;
	scanf("%d %d", &S, &N);
	int s[N], v[N];
	for(int i = 0; i < N; i++)
		scanf("%d %d", &s[i], &v[i]);
	printf("%d\n", solve(S, N - 1, s, v));
	return 0;
}
